### PR TITLE
[libc++][NFC] Centralize test for support of == and != in ranges

### PR DESCRIPTION
**Pull request by hawkinsw**
Previously, tests for whether comparison using == was supported by iterators derived from ranges adaptors was spread throughout the testing codebase. This PR centralizes the implementation of those tests.

### CODE REVIEW

**Comment by hawkinsw** [comment]
I noticed that these tests were implemented differently in different places throughout the range adapter test code. I hope that this is something that is useful!

**Comment by hawkinsw** [comment]
Hello @cjdb! Just let me know if there is anything I can do to make this better!

**Comment by hawkinsw** [comment]
Hope that this is helpful! Let me know if there is anything I can do to improve it! cc @var-const @cjdb

**Comment by hawkinsw** [comment]
I hope that this refactoring is useful!
Will

**Comment by hawkinsw** [comment]
> I hope that this refactoring is useful! Will

Please let me know if there is anything I can do to make this more useful.

File: libcxx/test/support/test_range.h
Comment (ldionne) [inline libcxx/test/support/test_range.h:94]: I would remove the default template argument since `weakly_equality_comparable_with<Iterator>` is a bit confusing. I think it increases clarity if we force writing `weakly_equality_comparable_with<Iterator, Iterator>`.

File: libcxx/test/support/test_range.h
Comment (ldionne) [inline libcxx/test/support/test_range.h:17]: We can't include private libc++ headers in the test suite since the test suite must work with other implementations too. Also I think this header is not necessary anyway.

**Review by ldionne — COMMENTED** [review]
This makes sense to me, with a few comments.

File: libcxx/test/support/test_range.h
Comment (var-const) [inline libcxx/test/support/test_range.h:94]: Optional: perhaps add a comment like `// [concept.equalitycomparable]` to indicate that this concept comes from the Standard (where it's exposition-only so we can't use it directly)?

File: libcxx/test/std/ranges/range.adaptors/range.elements/sentinel/equality.pass.cpp
Comment (var-const) [inline libcxx/test/std/ranges/range.adaptors/range.elements/sentinel/equality.pass.cpp:83]: Preexisting issue: unless I'm missing something, this repeats line 74 (71 after refactoring), `R` should probably be `CrossComparableR`.

File: libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp
Comment (var-const) [inline libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp:89]: Same -- `R` should probably be `CrossComparableR`.

**Review by var-const — CHANGES_REQUESTED** [review]
Nice refactoring, thanks! Mostly LGTM except a couple minor comments (in addition to the existing feedback).

File: libcxx/test/std/ranges/range.adaptors/range.elements/sentinel/equality.pass.cpp
Comment (hawkinsw) [reply libcxx/test/std/ranges/range.adaptors/range.elements/sentinel/equality.pass.cpp:83]: I concur. It looks like there is a pattern of four (4) tests for each of `R` and `CrossComparableR` (i.e.,

| left operand | right operand |
| -- | -- |
|  type | type |
| const type | type |
|type | const type |
|const type | const type |

and there was a simple copy/paste error here. I have addressed it with the update!

Good eyes!

File: libcxx/test/support/test_range.h
Comment (hawkinsw) [reply libcxx/test/support/test_range.h]: I believe that I addressed this with ee61f15cfef3fd8b08d818d7455c2192d94980ea.

Comment (hawkinsw) [reply libcxx/test/support/test_range.h]: I believe that I addressed this with ee61f15cfef3fd8b08d818d7455c2192d94980ea.

File: libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp
Comment (hawkinsw) [reply libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp]: I believe that I addressed this with ee61f15cfef3fd8b08d818d7455c2192d94980ea.

**Comment by hawkinsw** [comment]
Hello!

Just let me know if there is anything remaining that I can do to make this better!

File: libcxx/utils/libcxx/test/format.py
Comment (var-const) [inline libcxx/utils/libcxx/test/format.py:285]: Looks like a debugging left-over.

**Review by var-const — APPROVED** [review]
Thanks for addressing the feedback! LGTM (with one very straightforward comment).

File: libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp
Comment (var-const) [reply libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp]: Thanks! By the way, feel free to skip putting a link to the exact commit -- there's nothing wrong with doing that, of course, but I imagine it adds extra overhead for you to post these links that can easily be avoided (I'm looking at the whole delta since the last review anyway, and I would imagine other reviewers do so as well). Just trying to save you some time/friction!

Comment (hawkinsw) [reply libcxx/test/std/ranges/range.adaptors/range.take.while/sentinel/equality.pass.cpp]: Thank you! I appreciate that feedback! I wasn't sure how most reviewers do subsequent reviews.

**Comment by hawkinsw** [comment]
Because I had to rebase and fix a merge conflict, I just wanted to have a second pair of eyes on the updated PR before I squash/merge. Sorry for the additional bandwidth, @var-const !

File: libcxx/test/support/test_range.h
Comment (var-const) [inline libcxx/test/support/test_range.h:99]: Can you please readd the comment that shows this is a standard (exposition-only) concept?

**Review by var-const — APPROVED** [review]

**Comment by hawkinsw** [comment]
If everyone is okay with this (especially @var-const ), I am more than willing to Squash and Merge.

**Review by var-const — APPROVED** [review]

**Pull request merged**
